JIT crash for java/lang/StringConcatHelper.stringOf(Ljava/lang/Object;)Ljava/lang/String;

Java -version output
--------------------
openjdk version "17.0.10" 2024-01-16
IBM Semeru Runtime Open Edition 17.0.10.0 (build 17.0.10+7)
Eclipse OpenJ9 VM 17.0.10.0 (build openj9-0.43.0, JRE 17 Linux s390x-64-Bit Compressed References 20240116_630 (JIT enabled, AOT enabled)
OpenJ9   - 2c3d78b48
OMR      - ea8124dbc
JCL      - 2aad089841f based on jdk-17.0.10+7)

Summary of problem
------------------

Running the Hibernate ORM testsuite triggers a JVM crash. Also see https://ci.hibernate.org/job/hibernate-orm-pipeline/job/wip%2F7.0/20/execution/node/66/log/

The problem can be reproduced by running `./gradlew check -PexcludeTests=**/KotlinProjectTests* -Plog-test-progress=true --stacktrace` with the [`wip/7.0` branch of Hibernate ORM](https://github.com/hibernate/hibernate-orm/tree/wip/7.0).

Diagnostic files
----------------

[javacore.20240423.090841.2949679.0002.txt](https://github.com/eclipse-openj9/openj9/files/15078731/javacore.20240423.090841.2949679.0002.txt)
[jitdump.20240423.090841.2949679.0004.dmp](https://github.com/eclipse-openj9/openj9/files/15078733/jitdump.20240423.090841.2949679.0004.dmp)
[Snap.20240423.090841.2949679.0003.trc.txt](https://github.com/eclipse-openj9/openj9/files/15078736/Snap.20240423.090841.2949679.0003.trc.txt)

```
Unhandled exception
Type=Segmentation error vmState=0x00000000
J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
Handler1=000003FFAA6C9410 Handler2=000003FFAA5B19D8 InaccessibleAddress=0000000010000000
gpr0=0000000000191C00 gpr1=0000000010000000 gpr2=0000000010000000 gpr3=000003FF89AB91FA
gpr4=000003FF883CBEF4 gpr5=00000000023A83B8 gpr6=0000000010000000 gpr7=00000000319EA0C0
gpr8=0000000010000000 gpr9=000003FF8992DE00 gpr10=000003FF8992DDF4 gpr11=0000000000318F00
gpr12=00000000EC5C95A8 gpr13=000000000006D200 gpr14=000003FF8992DE84 gpr15=000003FFAB2F1BA0
psw=000003FF89AB9226 mask=0705200180000000 fpc=00080000 bea=000003FF8992DE82
fpr0 3f40000000000000 (f: 0.000000, d: 4.882812e-04)
fpr1 000003ffaaf9f968 (f: 2868508928.000000, d: 2.172219e-311)
fpr2 00000000001d28d8 (f: 1911000.000000, d: 9.441594e-318)
fpr3 0000000000000007 (f: 7.000000, d: 3.458460e-323)
fpr4 000003ffa4000a78 (f: 2751465984.000000, d: 2.172161e-311)
fpr5 000003ffa4000f70 (f: 2751467264.000000, d: 2.172161e-311)
fpr6 0000000000000000 (f: 0.000000, d: 0.000000e+00)
fpr7 9ae13b28b86404a3 (f: 3093562624.000000, d: -3.322044e-179)
fpr8 0000000030678b80 (f: 812092288.000000, d: 4.012269e-315)
fpr9 0000000002f39458 (f: 49517656.000000, d: 2.446497e-316)
fpr10 00000000306783a8 (f: 812090304.000000, d: 4.012259e-315)
fpr11 0000000000000000 (f: 0.000000, d: 0.000000e+00)
fpr12 000616c33b5900b9 (f: 995688640.000000, d: 8.467681e-309)
fpr13 000003fecc016a08 (f: 3422644736.000000, d: 2.170371e-311)
fpr14 0000000000040000 (f: 262144.000000, d: 1.295163e-318)
fpr15 000003fecc031588 (f: 3422754304.000000, d: 2.170371e-311)

Compiled_method=java/lang/StringConcatHelper.stringOf(Ljava/lang/Object;)Ljava/lang/String;
Target=2_90_20240116_630 (Linux 4.18.0-425.3.1.el8.s390x)
CPU=s390x (8 logical CPUs) (0x3ddbc6000 RAM)
----------- Stack Backtrace -----------
 (0x000003FF89AB9226 [<unknown>+0x0])
---------------------------------------
JVMDUMP039I Processing dump event "gpf", detail "" at 2024/04/23 09:08:41 - please wait.
JVMDUMP032I JVM requested System dump using '/home/linux1/workspace/hibernate-orm-pipeline_wip_7.0/hibernate-core/target/OOM-dump/core.20240423.090841.2949679.0001.dmp' in response to an event
JVMPORT030W /proc/sys/kernel/core_pattern setting "|/usr/lib/systemd/systemd-coredump %P %u %g %s %t %c %h %e" specifies that the core dump is to be piped to an external program.  Attempting to rename either core or core.2950261.  Review the manual for the external program to find where the core dump is written and ensure the program does not truncate it.

JVMPORT049I The core file created by child process with pid = 2950261 was not found. Review the documentation for the /proc/sys/kernel/core_pattern program "|/usr/lib/systemd/systemd-coredump %P %u %g %s %t %c %h %e" to find where the core file is written and ensure that program does not truncate it.

JVMDUMP012E Error in System dump: /home/linux1/workspace/hibernate-orm-pipeline_wip_7.0/hibernate-core/target/OOM-dump/core.20240423.090841.2949679.0001.dmp
JVMDUMP032I JVM requested Java dump using '/home/linux1/workspace/hibernate-orm-pipeline_wip_7.0/hibernate-core/target/OOM-dump/javacore.20240423.090841.2949679.0002.txt' in response to an event
JVMDUMP010I Java dump written to /home/linux1/workspace/hibernate-orm-pipeline_wip_7.0/hibernate-core/target/OOM-dump/javacore.20240423.090841.2949679.0002.txt
JVMDUMP032I JVM requested Snap dump using '/home/linux1/workspace/hibernate-orm-pipeline_wip_7.0/hibernate-core/target/OOM-dump/Snap.20240423.090841.2949679.0003.trc' in response to an event
JVMDUMP010I Snap dump written to /home/linux1/workspace/hibernate-orm-pipeline_wip_7.0/hibernate-core/target/OOM-dump/Snap.20240423.090841.2949679.0003.trc
JVMDUMP032I JVM requested JIT dump using '/home/linux1/workspace/hibernate-orm-pipeline_wip_7.0/hibernate-core/target/OOM-dump/jitdump.20240423.090841.2949679.0004.dmp' in response to an event
JVMDUMP051I JIT dump occurred in 'Test worker' thread 0x000000000006D200
JVMDUMP053I JIT dump is recompiling java/lang/StringConcatHelper.stringOf(Ljava/lang/Object;)Ljava/lang/String;
JVMDUMP013I Processed dump event "gpf", detail "".
```

@beikov  I am working on the issue on background and as issue is tagged fro 0.48 release, aiming to fix it by then. I do apologize it took bit long for me, but as I mentioned in https://github.com/eclipse-openj9/openj9/issues/19369#issuecomment-2183130152, with recent changes that went in hibernate I could not reproduce the failure as I started hitting following error 
```
* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':hibernate-jfr:compileJava'.
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:130)
	at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:282)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:128)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:116)
...
Caused by: org.gradle.api.internal.tasks.compile.CompilationFailedException: Compilation failed; see the compiler error output for details.
	at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:61)
	at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:41)
```

Which I saw on x86 linux as well. I had to find the commit to roll back reproduce this and continue investigation.

--- a/runtime/compiler/ilgen/Walker.cpp
+++ b/runtime/compiler/ilgen/Walker.cpp
@@ -1630,9 +1630,6 @@ TR_J9ByteCodeIlGenerator::stashArgumentsForOSR(TR_J9ByteCode byteCode)
 
    TR::MethodSymbol *symbol = symRef->getSymbol()->castToMethodSymbol();
    int32_t numArgs = symbol->getMethod()->numberOfExplicitParameters() + (symbol->isStatic() ? 0 : 1);
-   // For OpenJDK MH implementation, some args for invokedynamic/invokehandle (details below)
-   // that are already pushed to stack must not be stashed
-   int32_t numArgsToNotStash = 0;
 
 #if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
    // If the transition target is invokehandle/invokedynamic, the arguments to be
@@ -1671,6 +1668,7 @@ TR_J9ByteCodeIlGenerator::stashArgumentsForOSR(TR_J9ByteCode byteCode)
    // Notice that we always generate a resolved call, thus we use unresolvedInCP to tell
    // us whether the side table entry is resolved
    //
+   int32_t numArgsToNotStash = 0;
    if (byteCode == J9BCinvokedynamic ||
        byteCode == J9BCinvokehandle)
       {
@@ -1684,10 +1682,10 @@ TR_J9ByteCodeIlGenerator::stashArgumentsForOSR(TR_J9ByteCode byteCode)
       if (trace())
          traceMsg(comp(), "Original num args for invokedynamic/handle: %d, num args to not stash for OSR: %d, stack size: %d\n", numArgs, numArgsToNotStash, _stack->size());
       }
-#endif
+
    numArgs -= numArgsToNotStash;
-   // For OpenJDK MethodHandle implementation, there may be items on stack that we need to exclude
-   int32_t adjustedStackSize = _stack->size() - numArgsToNotStash;
+#endif
+
    TR_OSRMethodData *osrMethodData =
       comp()->getOSRCompilationData()->findOrCreateOSRMethodData(comp()->getCurrentInlinedSiteIndex(), _methodSymbol);
    osrMethodData->ensureArgInfoAt(_bcIndex, numArgs);
@@ -1696,10 +1694,10 @@ TR_J9ByteCodeIlGenerator::stashArgumentsForOSR(TR_J9ByteCode byteCode)
    // It is necessary to walk the whole stack to determine the slot numbers
    int32_t slot = 0;
    int arg = 0;
-   for (int32_t i = 0; i < adjustedStackSize; ++i)
+   for (int32_t i = 0; i < _stack->size(); ++i)
       {
       TR::Node * n = _stack->element(i);
-      if (adjustedStackSize - numArgs <= i)
+      if (_stack->size() - numArgs <= i)
          {
          TR::SymbolReference * symRef = symRefTab()->findOrCreatePendingPushTemporary(_methodSymbol, slot, getDataType(n));
          osrMethodData->addArgInfo(_bcIndex, arg, symRef->getReferenceNumber());

